File: TopInterview150/Backtracking/GenerateParentheses/solution.py
class Solution(object):
    def backtrack(self, result, comb, n, open, close):
        if (len(comb) == n*2):
            result.append(comb[:])
            return
        
        if open < n:
            self.backtrack(result, comb + '(', n, open+1, close)

        if close < open:
            self.backtrack(result, comb + ')', n, open, close+1)
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        comb = ""
        self.backtrack(result, comb, n, 0, 0)
        return result

File: TopInterview150/Backtracking/GenerateParentheses/solution.cpp
class Solution {
    public:
        vector<string> generateParenthesis(int n) {
            vector<string> result;
            string comb;

            backtrack(result, comb, n, 0, 0);
            return result;

        }

        void backtrack(
            vector<string>& result,
            string& comb,
            int n,
            int open,
            int close
        )
        {
            if (n*2 == comb.size())
            {
                result.push_back(comb);
                return;
            }

            if (open < n)
            {
                comb.push_back('(');
                backtrack(result, comb, n, open + 1, close);
                comb.pop_back();
            }

            if (close < open) 
            {
                comb.push_back(')');
                backtrack(result, comb, n, open, close + 1);
                comb.pop_back();
            }
        }
    };